### PR TITLE
AI cell generation

### PR DESCRIPTION
**Pull request by nichochar**
Introduce the ability to create cells (code, markdown) with AI.

Below is a video of the MVP experience. A few things to note:
* I built it so that it can create either markdown or code. There was little in the design that prevented this, and technically it was natural. 
* We need to figure out how to give more context & data when things go wrong. Our users are developers so they'll immediately want to see logs, that's at least what I want.

https://www.loom.com/share/91e4905a691f424db2791539ddfaf3e8?sid=683873e2-0fae-4f98-8019-063002ff3b47

### CODE REVIEW

File: packages/web/src/components/cells/generate-ai.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+        <div className="flex items-start">
+          <Sparkles size={16} className="m-2.5" />
+
+          <textarea
+            value={prompt}
+            autoFocus
+            onChange={(e) => setPrompt(e.target.value)}
+            placeholder="Write a prompt..."
+            className="flex min-h-[60px] w-full rounded-sm px-3 py-2 text-sm placeholder:text-muted-foreground focus-visible:outline-none pl-0"
```
Comment (benjreinhart) [inline packages/web/src/components/cells/generate-ai.tsx:103]: Why not codemirror here so we can get auto-resize? Otherwise, we should make this a bit bigger and add a `resize-none`

File: packages/api/server/http.mts [hunk 2]
```diff
@@ -128,6 +128,25 @@ router.post('/generate', cors(), async (req, res) => {
   }
 });
 
+// Generate a cell using AI from a query string
+router.options('/sessions/:id/generate_cell', cors());
```
Comment (benjreinhart) [inline packages/api/server/http.mts:132]: Why not a websocket event?

File: packages/api/prompts/cell-generator-javascript.txt [hunk 3]
```diff
@@ -0,0 +1,128 @@
+===== BEGIN INSTRUCTIONS CONTEXT =====
+
+You are tasked with generating a Srcbook code cell for the user.
+
+A Srcbook is a JavaScript notebook following a markdown-compatible format called .srcmd. It's an interactive and rich way of programming that follows the literate programming idea.
```
Comment (benjreinhart) [inline packages/api/prompts/cell-generator-javascript.txt:5]: >  It's an interactive and rich way of programming that follows the literate programming idea.

These prompts feel too wordy IMO. This ^, for example, doesn't seem necessary.

**Review by benjreinhart — APPROVED** [review]
IMO, the prompts look too long and wordy to me. There seems to be a lot of unnecessary text in them (why do we need that lengthy example?).

File: packages/api/server/http.mts (hunk 2 above)
Comment (nichochar) [reply packages/api/server/http.mts:132]: I don't have a super strong reason: I favor http when i need request/response semantics with a pending state. It's also less code.

Do you think this would be better as websocket?

File: packages/api/prompts/cell-generator-javascript.txt (hunk 3 above)
Comment (nichochar) [reply packages/api/prompts/cell-generator-javascript.txt:5]: Yes, the prompts can definitely be iterated on. This made more sense in the generate srcbook one...I'll remove that particular sentence

Without evals, it's a bit of a nacked process. I am happy to remove this, but the next step for prompt iteration is to set up some basic evals. For this particular flow, the evals could programmatically test that we get valid srcbook code back based on a set of known user request examples. 


Once we have that, we can start iterating with a bit more confidence.

File: packages/web/src/components/cells/generate-ai.tsx (hunk 1 above)
Comment (nichochar) [reply packages/web/src/components/cells/generate-ai.tsx:103]: > Why not codemirror here so we can get auto-resize?

Codemirror felt heavyweight for what is essentially a simple multi-line input

> we should make this a bit bigger and add a resize-none

I like the resize, it's not too visible, but gives the user control should they be pasting a bigger prompt (unlikely, but possible). Why don't you like it?

**Comment by nichochar** [comment]
> IMO, the prompts look too long and wordy to me. There seems to be a lot of unnecessary text in them (why do we need that lengthy example?).

Do you have data that suggests that shorter prompts work better? I am happy to iterate on the prompts and shorten them, but I think that the example is one of the most valuable parts. LLMs do very well when you give them examples (few shot).

Are your concerns on cost? latency? 
I'm meeting with the codestory people to talk about this stuff, but I don't have a lot of intuitions yet. We need to setup evals and experiment. This is a first pass which does quite well on my empirical QA.
